2025-05-99-how-to-implement-zero-copy-communication/index.md

```diff
@@ -0,0 +1,164 @@
+No copies are involved after the data is created.
+
+<table width=100% height=100% border=0><tr><td align=center>
+
+<img src="shared-memory-idea.png" alt="shared-memory-idea" style="max-width: 50%"/>
+
+</td></tr></table>
+
+### Hands-On: iceoryx2 in Action
+
+We’ll now walk through the
+[iceoryx2 publish-subscribe example](https://github.com/eclipse-iceoryx/iceoryx2/tree/main/examples/rust/publish_subscribe),
+where data is sent by a publisher to multiple subscribers using zero-copy.
+
+Sensor APIs often provide a pointer to the latest data:
+
+```rust
+let data_ptr = sensor_get_next_frame();
+```
+
+However, this is not zero-copy: the data resides in sensor-managed memory, and
+must be copied into shared memory before sending. For true zero-copy, we
+need the sensor to write directly into shared memory, provided by the
+zero-copy communication framework.
+
+* **1. Produce Data into Shared Memory**
+  First, we acquire an uninitialized sample from the publisher. This is a chunk
+  of memory from the publisher’s data segment, reserved for the user's payload.
+  Under the hood, an allocator manages this memory block:
+
+  ```rust
+  let mut uninitialized_sample =
+      publisher.loan_slice_uninit(MAX_SIZE_OF_SENSOR_DATA_FRAME)?;
+  ```
+
+  Next, the sensor writes directly into this memory:
+
+  ```rust
+  sensor_get_next_frame(uninitialized_sample.payload_mut());
+  ```
+
+  * **Deep Dive:**
+    When the publisher is created:
+    1. A shared memory object is created using `shm_open`.
+    2. It is resized via `ftruncate` to reserve enough space.
+    3. The segment is mapped into the process using `mmap`.<br><br>
+
+    In iceoryx2 the memory of the data segment is managed by a pool allocator.
+    It partitions memory into fixed-size regions, which avoids fragmentation
+    and allows predictable allocation times - essential for real-time and
+    safety-critical systems.
+
+    The call `loan` allocates a chunk of memory from this pool allocator and
+    returns it to the user.
+
+* **2. Send Pointer To Data**
+  Once the data is produced, the sample is marked as initialized and published:
+
+  ```rust
+  let sample = unsafe { uninitialized_sample.assume_init(); }
+  sample.send()?;
+  ```
+
+  * **Deep Dive:**
+    Only an offset to the memory location is sent to the subscribers, not the
+    actual data.
+    Each process has its own virtual memory layout, so we cannot directly
+    share pointers. Instead:
+    * The publisher calculates an offset from the base of the shared memory segment.
+    * The subscriber adds this offset to its local mapping of that segment.
+    * A valid pointer to the received data is reconstructed.<br><br>
+
+    To transfer the offset, an inter-process communication mechanisms like
+    message queues, pipes, unix domain sockets, or shared memory queues can be
+    used.
+
+    <table width=100% height=100% border=0><tr><td align=center>
+
+    <img src="offset.png" alt="offset" style="max-width: 80%"/>
```

Comment:
It took me some time to understand the image. What do you think of something like this

```
                Shared
          ^     Memory          ^ 
          |                     |
   0x4000 | +------------*      | 0xD000
          | |            |      |
          | +------------*      |
          | | payload    | ---  |
          | +------------*  |   |
          | |            |  |   |
   0x3000 | +------------* ---  | 0xC000
          |              offset |
          |                     |
          |                     |
         ---                   ---
     Address Space         Address Space
       Publisher             Subscriber
```